src/withdrawal_limit/component.cairo

```diff
@@ -0,0 +1,126 @@
+        // in this day (if the value is x, the amount left to withdraw is x-1). 0 means that
+        // currently there was no withdrawal from this token in this day or there were withdrawals
+        // but the limit flag was turned off.
+        remaining_intraday_withdraw_quota: LegacyMap<(ContractAddress, u64), u256>,
+        // The daily withdrawal limit percentage.
+        daily_withdrawal_limit_pct: u8,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    pub enum Event {
+        RemainingQuotaUpdated: RemainingQuotaUpdated
+    }
+
+    #[derive(Drop, starknet::Event)]
+    pub struct RemainingQuotaUpdated {
+        new_quota: u256
+    }
+
+
+    #[embeddable_as(WithdrawalLimitImpl)]
+    pub impl WithdrawalLimit<
+        TContractState, +HasComponent<TContractState>, +ITokenBridge<TContractState>
+    > of IWithdrawalLimit<ComponentState<TContractState>> {
+        // Returns the current remaining withdrawal quota for a given token. If there is no limit,
+        // returns max uint256. If the limit was not set yet, we calculate it based on the total
+        // supply. Otherwise, return the limit.
+        fn get_remaining_withdrawal_quota(
+            self: @ComponentState<TContractState>, token: ContractAddress
+        ) -> u256 {
+            // If there is no limt, return max uint256.
+            if self.get_contract().is_withdrawal_limit_applied(:token) == false {
+                return BoundedInt::max();
+            }
+            let remaining_quota = self.read_withdrawal_quota_slot(:token);
+            if remaining_quota == 0 {
+                return self.get_daily_withdrawal_limit(:token);
+            }
+            remaining_quota - constants::REMAINING_QUOTA_OFFSET
+        }
+    }
+
+    #[generate_trait]
+    pub impl InternalImpl<
+        TContractState, +HasComponent<TContractState>, +ITokenBridge<TContractState>
+    > of InternalTrait<TContractState> {
+        // Sets the remaining withdrawal quota for today.
+        fn set_remaining_withdrawal_quota(
+            ref self: ComponentState<TContractState>, token: ContractAddress, amount: u256
+        ) {
+            let now = get_block_timestamp();
+            let day = now / constants::SECONDS_IN_DAY;
+            self
+                .remaining_intraday_withdraw_quota
+                .write((token, day), amount + constants::REMAINING_QUOTA_OFFSET);
+        }
+
+        // Returns the remaining withdrawal quota for today.
+        fn read_withdrawal_quota_slot(
+            self: @ComponentState<TContractState>, token: ContractAddress
+        ) -> u256 {
+            let now = get_block_timestamp();
+            let day = now / constants::SECONDS_IN_DAY;
+            self.remaining_intraday_withdraw_quota.read((token, day))
+        }
+
+
+        // Try to withdraw an amount and if it succeeds, update the remaining withdrawal quota.
+        fn consume_withdrawal_quota(
```

Comment:
we're not using this anywhere rn, is it a TODO?

Reply:
Yes, this part is still to implement. Will work on this in upcoming prs